### ISSUE TITLE
Using a class as return type of function will accept a return an object contructed with `{}`

### ISSUE THREAD

**Issue by diasbruno** [microsoft/TypeScript#52765]
# Bug Report

Declaring a class `A` with only properties like `b: string` and `c: number`,
and used as a return type of a function that returns a object contructed with `{}`,
the type system will accept this return type as a valid instance of `A`.

Adding a method to class `A` will cause the types to misalign...

Is this behavior correct? Both objects are equivalent, 
but their types/prototypes are differente.

### 🔎 Search Terms

object, type, class, type inference, type coersion

### 🕗 Version & Regression Information

Seems to be the case for all versions. v4.9.5

### ⏯ Playground Link

[Playground](https://www.typescriptlang.org/play?ts=4.9.5#code/MYGwhgzhAECC0G8BQ1XQEYC5oQC4CcBLAOwHNoBeaAImpTWAHti98BXYXR-ACixwIlSASkT00aXAAtCEAHTpKGANzjoAXySakAMzbFOhZtAAePYdnjIJ+AKa42+YogzZqjANY0Nq7aEgwAELW0MDYxGwAtui2+EoAjKoSqPSk9tCB5i52Dk7Q0rJywMoaWki6+obGAJ7m2IFiNvaOzgih2PE+WkA)

### 💻 Code

```ts
class A {
    b: string = ""
    constructor(b: string) {
        this.b = b;
    }
}

function x(): A {
    return { b: "ok " }; // compile just fine
}

class B{
  c: number = 1;
      
  get getter() { return this.c; }
}

function y(): B {
    return { c: 1 };
    // Property 'getter' is missing in type '{ c: number; }' but required in type 'B'.
}
```

### 🙁 Actual behavior

Using a class as return type should allow an raw object.

### 🙂 Expected behavior

This behavior shouldn't be allowed because they are different instances.

**Comment by RyanCavanaugh** [comment]
Classes with only public members are structurally typed. This is the intended behavior.

**Comment by fatcerberus** [comment]
Workaround: Ensure the class has at least one private property.

**Comment by diasbruno** [comment]
Thanks, @RyanCavanaugh. I saw this on a code snippet that someone forgot to instantiate a class, something like

```ts
function x(): A {
  return new A({ ... props of A }); 
}
```

and I was currious to see if it is a bug or feature.

Another question should we consider this an implicit cast? Seems a little odd, but ok...

**Comment by diasbruno** [comment]
I mean, it says that this should be equal

```ts
{ b: number; } == class { b: number }
```

**Comment by diasbruno** [comment]
@fatcerberus it would be hard to see a class without any methods, but it could, at least, guarantee better packing of the data or something (I guess)...

**Comment by RyanCavanaugh** [comment]
>  I saw this on a code snippet that someone forgot to instantiate a class

The sample has an instantiation. Can you elaborate?

**Comment by diasbruno** [comment]
Sure. This is what should be the correct version of the snippet

```ts
interface PropsOfA {}

class A {
  constructor(props: PropsOfA) {}
}

function factory(props: Partial<PropsOfA>) => A {
  return new A({ ...props });
}
```

but on the `factory()` the person forgot to instantiating `A` leaving just the object.

```ts
function factory(props: Partial<PropsOfA>) => A {
  return { ...props };
}
```

**Comment by RyanCavanaugh** [comment]
Right, allowing that is the intended behavior due to the structural match

**Comment by Crinax** [comment]
I want to open this issue, because the class instance type is not equals to object type with same fields.

For example sandbox: https://www.typescriptlang.org/play?ts=5.4.3#code/MYGwhgzhAECC0G8BQ1XWAewHYQC4CcBXYXDfACgEsATALmj30qwHMBKRFNb3AC0ogA6GtAC80GgG4uqAL4zoCmvUbMW0+Uma4ApvgBmYYDugBJeMm7KGBNRqRJ9hLCUrZoAWzABrHbHIADvgYARD05mz0Fgr4OriE+Fic3FZ00EEhQjQKsvZImDi4nmAAngBGfmLFvv4IEmkA5JDU+g3QsmzSWvrQ5F7llcx4YC46GD2wHJZoBRAYIDqCIBgs5A0CcA2dSJqUPX2lFfBDuCPG49AA8mUAVjokUwqz84vLq+sw13ckWxpAA

I have an example of strange logic: we can define two same classes (A and B), and return class B from function that have return type A

Sandbox: https://www.typescriptlang.org/play?ts=5.4.3#code/MYGwhgzhAECC0G8BQ1XWAewHYQC4CcBXYXDfACgEsATALmj30qwHMBKRFNb3AC0ogA6GtAC80GgG4uqAL4zoCmvUbMW0+UlCQYAIU7dMOAsVIVlDAmo7JuPfkJHipC+dyV1LTVhqRJmuACm+ABmYMCB0ACS8LZoFqo+SJohhFgklNjQALZgANaBsOQADvgYxRD0MWz0sQr4gbiE+FjQWIEA7tC6JWUVwtRsvlrYeDlgAJ4ARoVi4wVFCBKeAOSQ1CEr0LJDfpQh0OS507PMeGDpgRgHsDYKRhAYIIGCIBgs5CsCcCu7mvuHY4zeBnXAXCLXbp3QyjJ4vN4fL56X7DAFHSbAiTGcFXA4AeSmACtAiRoWgHnDXu9Pt8CcSSCjkkA

In this examples we cannot check type in runtime correctly
